VM Affinity (#383)
openshift/openshift-origin-design: designs/virtualization/4.5/Scheduling/Affinity/Affinity.md

---
parent: Virtualization
version: 4.5
---

# Affinity

Introducing Affinity as it appears in-page and the Edit modal.

## in-page placement

> ![Affinity field in the Scheduling section ](img/Affinity-0-0.jpg)

Reviewer: nit: can we change these designs to virtualization in the nav :)

> The affinity field is placed in the Workload details page, Details tab, Scheduling section, Policies columns.

Reviewer: ```suggestion
The affinity field is placed in the Workload details page > Details tab > Scheduling section > Policies columns.
```

![Affinity field in the Scheduling section ](img/Affinity-0-1.jpg)
The affinity field counts the number of rules applied.

## Affinity rules list

> ![Affinity all rules modal ](img/Affinity-1-0.jpg)

Reviewer: Rules with "Preferred" - with instead of will

Reviewer: Should the title of this modal be "Affinity rules" or "affinity rules overview". It feels a bit odd to have two edit modals.

Author: Right, will change to "Overview".
"List" is too technical and obvious, right?

Reviewer: I suppose it could be list but I prefer Overview :) But don't feel strongly.

Author: Overview it is

> On clicking on the counter Or the Edit pencil icon, the user will be presented with a modal, containing all the applied rules for this workload.

Reviewer: ```suggestion
On clicking on the counter or the Edit pencil icon, the user will be presented with a modal, containing all the applied rules for this workload.
```

Reviewer: maybe I'm missing something, why are we adding both a pencil and the counter as a way to get to the modal? Wasn't this changed to just the pencil?

Author: I don't recall dropping the clickability of the counter. But I will remove it so it won't create confusion.

Clicking on the 'Add affinity rule' button will replace this modal with the 'Add affinity rule' modal.
Selecting 'Edit' from each existing rule kebab menu will replace this modal with the 'Edit affinity rule' modal

A support text below the header will notify the user that:

> "Rules will ‘Preferred’ condition will stack with an ‘AND’ relation between them. 

Reviewer: ```suggestion
"Rules with ‘Preferred’ condition will stack with an ‘AND’ relation between them. 
```

Rules with ‘Required condition will have an ‘OR’ relation between them."

## Add/Edit affinity rule modal

![Affinity Edit/Create modal ](img/Affinity-2-0.jpg)

> TThe default fields for an affinity rule are:

Reviewer: ```suggestion
The default fields for an affinity rule are:
```

- type: Node affinity

> - Condtion: Preferred

Reviewer: ```suggestion
- Condition: Preferred
```

> - Weight: 1 (out of 100)

Reviewer: is it common to know the limit is 100? Should we add "(of 100)" to the label?

Author: I've added it to the support text just below the field. I also added a "Lowest" and "Highest" indication to explain the scale to the user. 
Do u think it is not prominent enough? By label do you mean the field title?

Reviewer: ya this is better than my idea :)
Maybe it could read 
`Value should between 1(lowest) and 100(highest)`

- Select nodes via: Node labels
- Operator (between Key and Value): In

There is no need to mention the Default to the user.

![Affinity Edit/Create modal - Required ](img/Affinity-2-1.jpg)
Switching to 'Required during scheduling' option from the Condition field, the Weight field will be excluded.

![Affinity Edit/Create modal - Any expressions ](img/Affinity-2-2.jpg)

> Swtiching to 'Any key:value expression on YAML' will change the Key field to be able to accept stings. The placeolder will reflect that as well.

Reviewer: ```suggestion
Switching to 'Any key:value expression on YAML' will change the Key field to be able to accept strings. The placeholder will reflect that as well.
```

'Any key:value expression on YAML' option will also change the 'Add button' from 'Add label' to 'Add expression'.

![Affinity Edit/Create modal - Any expressions ](img/Affinity-2-3.jpg)
On 'Any expression' option, the user will be required to enter full path to the key:value pair.
The 'values' field is a multi-value field and will be applied using the white labels.

> Entering values in the label/expression rows ill relulst in a 'Matching nodes' notification.

Reviewer: ```suggestion
Entering values in the label/expression rows will result in a 'Matching nodes' notification.
```

![Affinity Edit/Create modal - Operator options ](img/Affinity-2-4.jpg)
When selecting the target node/workload to apply "affinity to"
the user can select one of the following Operator options:
- In (any node/workload with the specified key and has one or more of the following values)
- NotIn (any node/workload with the specified key none of the following labels)
- Exists (ny node/workload with the specified key - Values field is disabled)

> - DoesNotExost (any node/workload that does not have the specified key - Values field is disabled)

Reviewer: ```suggestion
- DoesNotExist (any node/workload that does not have the specified key - Values field is disabled)
```

---
parent: Virtualization
version: 4.5
---

# Affinity

> Introducing Affinity as it appears in-page and the Edit modal.

Reviewer: ```suggestion
Introducing Affinity as it appears in-page and the Edit/Add modal.
```

## in-page placement

> ![Affinity field in the Scheduling section ](img/Affinity-0-0.jpg)

Reviewer: in the implementation the text is not blue for any of the values of the sections

> The affinity field is placed in the Workload details page > Details tab > Scheduling section > Policies columns.

Reviewer: ```suggestion
Affinity is placed in the Workload details page > Details tab > Scheduling and Resource Requirements section > Policies column.
```

![Affinity field in the Scheduling section ](img/Affinity-0-1.jpg)
The affinity field counts the number of rules applied.

## Affinity rules list

> ![Affinity all rules modal ](img/Affinity-1-0.jpg)

Reviewer 2: Should this be "View affinity rules" ? And I'm not sure we use "Done" as the primary action for modals anywhere else. If users can delete rules from here, should the actions actually be save and cancel?

Author: Sounds right, I'll update it to Save/Cancel

> ![Affinity all rules modal ](img/Affinity-1-0.jpg)

Reviewer: In the implementation the action is "Add affinity" it drops the "rule"

> On clicking on the Edit pencil icon, the user will be presented with a modal, containing all the applied rules for this workload.

Reviewer: ```suggestion
When clicking on the Edit pencil icon, the user will be presented with a modal, containing all the applied rules for this workload.
```

Clicking on the 'Add affinity rule' button will replace this modal with the 'Add affinity rule' modal.
Selecting 'Edit' from each existing rule kebab menu will replace this modal with the 'Edit affinity rule' modal

A support text below the header will notify the user that:
"Rules with ‘Preferred’ condition will stack with an ‘AND’ relation between them. 
Rules with ‘Required condition will have an ‘OR’ relation between them."

The "learn more about affinity rules" links to:
https://kubevirt.io/2020/Advanced-scheduling-with-affinity-rules.html

## Add/Edit affinity rule modal

> ![Affinity Edit/Create modal ](img/Affinity-2-0.jpg)

Reviewer: affinity is spelt wrong under the Rule heading :)

Reviewer: looks like it's for all these images

> ![Affinity Edit/Create modal ](img/Affinity-2-0.jpg)

Reviewer: In the implementation the title for the modal is either "New affinity" or "Edit affinity" the action is "Save affinity" for both the add and edit modals.

The default fields for the following fields are:
- type: Node affinity

> - Condition: Preferredd

Reviewer: ```suggestion
- Condition: Preferred
```

- Weight: 1 (out of 100)
- Expression type: Label
- Operator (between Key and Value): In

There is no need to mention the Default to the user.

> ![Affinity Edit/Create modal - Required ](img/Affinity-2-1.jpg)

Reviewer: There are inconsistencies with what's been implemented versus designed. That's pretty normal, but we should update these design to reflect what's in the UI to avoid confusion
<img width="807" alt="Screen Shot 2020-05-08 at 11 47 03 AM" src="https://user-images.githubusercontent.com/287880/81417619-d57e6480-9121-11ea-8cae-a1e960514868.png">
